ffi/freetype.lua

```diff
@@ -186,18 +215,71 @@ end
 
 local FTFaceType = ffi.metatype("struct FT_FaceRec_", FTFace_mt) -- luacheck: ignore 211
 
+local FTSize_mt = {__index={}}
+
+function FTSize_mt.__index:checkGlyph(char)
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    return self.face:checkGlyph(char)
+end
+
+function FTSize_mt.__index:renderGlyph(char, bold)
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    local glyph = self.face:renderGlyph(char, bold)
+    return glyph
+end
+
+function FTSize_mt.__index:renderGlyphByIndex(index, embolden_half_strength)
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    return self.face:renderGlyphByIndex(index, embolden_half_strength)
+end
+
+function FTSize_mt.__index:getEmboldenHalfStrength(factor)
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    return self.face:getEmboldenHalfStrength(factor)
+end
+
+function FTSize_mt.__index:getHeightAndAscender()
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    return self.face:getHeightAndAscender()
+end
+
+function FTSize_mt.__index:getInfo()
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    return self.face:getInfo()
+end
+
+function FTSize_mt.__index:getKerning(leftcharcode, rightcharcode)
+    assert(ft2.FT_Activate_Size(self) == 0, "failed to activate font size")
+    return self.face:getKerning(leftcharcode, rightcharcode)
+end
+
+function FTSize_mt.__index:done()
+    ffi.gc(self, nil)
+    done_size(self)
+end
+
+local FTSizeType = ffi.metatype("struct FT_SizeRec_", FTSize_mt) -- luacheck: ignore 211
+
+local faces_cache = setmetatable({}, {__mode="v"})
+
 function FT.newFace(filename, pxsize, faceindex)
     if pxsize == nil then pxsize = 16*64 end
     if faceindex == nil then faceindex = 0 end
-    local face = new_face(filename, faceindex)
-    if ft2.FT_Set_Pixel_Sizes(face, 0, pxsize) ~= 0 then
-        face:done()
+    local cache_key = string.format('%s:%d', filename, faceindex)
+    local face = faces_cache[cache_key]
+    if not face then
+        face = new_face(filename, faceindex)
+        faces_cache[cache_key] = face
+    end
+    local size = new_size(face)
+    if ft2.FT_Activate_Size(size) ~= 0 or ft2.FT_Set_Pixel_Sizes(face, 0, pxsize) ~= 0 then
```

Comment:
I assume this means that we first try to activate it, and if this fails, it means it has not yet been "created/sized" (?) and that we need to call FT_Set_Pixel_Sizes() to "create/size" it and that this implcitely activate it ?
Some small comment to explain that gymnastic (and/or 2 distincts branches to make that clearer?) would be nice.

And yes, as my review comment on xtext.cpp, some small comment explaining what "size" is here too would be nice.